bundles/org.openhab.binding.doorbird/src/main/java/org/openhab/binding/doorbird/internal/DoorbirdUdpListener.java

```diff
@@ -0,0 +1,151 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.doorbird.internal;
+
+import java.io.IOException;
+import java.net.DatagramPacket;
+import java.net.DatagramSocket;
+import java.net.SocketException;
+import java.net.SocketTimeoutException;
+import java.util.Arrays;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DoorbirdUdpListener} is responsible for receiving
+ * UDP braodcasts from the Doorbird doorbell.
+ *
+ * @author Mark Hilbush - Initial contribution
+ */
+@NonNullByDefault
+public class DoorbirdUdpListener extends Thread {
+    // Doorbird devices report status on a UDP port
+    private static final int UDP_PORT = 6524;
+
+    // How long to wait in milliseconds for a UDP packet
+    private static final int SOCKET_TIMEOUT = 3000;
+
+    private static final int BUFFER_SIZE = 80;
+
+    private final Logger logger = LoggerFactory.getLogger(DoorbirdUdpListener.class);
+
+    private final DoorbirdEvent event = new DoorbirdEvent();
+
+    // Used for callbacks to handler
+    private DoorbirdHandler thingHandler;
+
+    // UDP socket used to receive status events from doorbell
+    private @Nullable DatagramSocket socket;
+
+    private byte @Nullable [] lastData;
+    private int lastDataLength;
+    private long lastDataTime;
+
+    public DoorbirdUdpListener(DoorbirdHandler thingHandler) {
+        this.thingHandler = thingHandler;
+    }
+
+    @Override
+    public void run() {
+        receivePackets();
+    }
+
+    public void shutdown() {
```

Comment:
I think there is a racing possible on socket so that main thread will have disposed of it and this thread will still use it. Maybe an exception will handle it, but it is not a graceful shutdown. This thread could handle interrupted exception and check for interrupted status instead of for socket to be null and close the socket. Main thread could only cancel(true) the job.

Reply:
I'll look into this.

Reply:
I remember now... Calling `cancel` won't interrupt the receive call on the UDP socket. We ran into this a while back on the [miio binding](https://community.openhab.org/t/xiaomi-robot-vacuum-binding/31317/462?u=mhilbush). If you still think there's an issue here, we can discuss further.

https://stackoverflow.com/questions/4670664/interrupt-a-thread-in-datagramsocket-receive

https://stackoverflow.com/questions/40055175/how-to-interrupt-a-blocking-call-to-udp-sockets-receive

https://stackoverflow.com/questions/7962121/how-can-i-stop-the-block-method-datagramsocket-receive-in-a-thread

Reply:
Oh, great to learn new things. I think if the only reason for IOException is closing the thread, this is good. But if IOException can occur for another reason, which I don't know, the racing is still possible and maybe a critical section on access to socket would be needed. Additionally, if it occurred, the thread would just wind down silently.